### PR TITLE
[jit] Compute the instance size/alignment correctly for gshared types whose constraint is a generic valuetype.

### LINKED ISSUE #18455: [2019-12] System.Core tests crash on iOS 64b devices.
In the integration PR (https://github.com/xamarin/xamarin-macios/pull/7537) we are running into issues when running the System.Core tests from the second part of the split test dll.

An example of the segfault can be found here:

07:44:26.2262310 	0x105a7e334 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : mono_dump_native_crash_info
07:44:26.2262380 	0x105a746bc - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : mono_handle_native_crash
07:44:26.2262670 	0x105a82a14 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : mono_sigsegv_signal_handler
07:44:26.2262730 	0x194880894 - /usr/lib/system/libsystem_platform.dylib : <redacted>
07:44:26.2262760 	0x105bb88c4 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : sgen_check_whole_heap_stw
07:44:26.2263920 	0x105bb7a04 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : sgen_check_whole_heap_stw
07:44:26.2264090 	0x105bb432c - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : sgen_perform_collection
07:44:26.2264420 	0x105bb40f4 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : sgen_ensure_free_space
07:44:26.2265700 	0x105baa2d4 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : sgen_alloc_obj_nolock
07:44:26.2265940 	0x105b63604 - /private/var/containers/Bundle/Application/A4488A3B-0BB2-4955-A96F-F648A0C80E5E/com.xamarin.bcltests.SystemCoreXunit Part 2.app/Frameworks/Mono.framework/Mono : mono_gc_alloc_vector

the tests only fail/crash on release mode on a device. The following links contain useful information:

* Full test run [report](http://xamarin-storage/jenkins/xamarin-macios/pull-request/6f9193f5775639a80101bd2bbcf2605aed544719/3382850/device-tests/jenkins-results/tests/index.html).
* Full [crash](http://xamarin-storage/jenkins/xamarin-macios/pull-request/6f9193f5775639a80101bd2bbcf2605aed544719/3382850/device-tests/jenkins-results/tests/%5bxUnit%5d%20Mono%20SystemCoreXunit%20Part%202/1853/run-dfa7a72864912a924fe96ee65ce8551f3210a068-20200114_074413.log).

### PR DESCRIPTION
**Pull request by vargaz**
Fixes #18455

### CODE REVIEW

**Comment by BrzVlad** [comment]
This fix doesn't seem to be complete. It fixes #18455 but doesn't seem to fix `System.Tests.ValueTupleTests.ValueTupleSerialization` from #18274, which was caused by the same llvm optimization. According to @lewurm, we don't have a clear isolated test case there so it might be a bit more painful to debug.

**Comment by vargaz** [comment]
What is the exact failure for these tests ?

**Comment by lewurm** [comment]
This assert fails: https://github.com/mono/corefx/blob/7c9e2158925339a5a96ff8d09d21c600a0b27e5a/src/System.ValueTuple/tests/ValueTupleTests.cs#L1232

**Comment by lewurm** [comment]
@imhameed and I failed yesterday to extract a repro for the `ThreeTuples ()` test case, but I was able to extract one from `ValueTupleSerialization ()` now. It fails on iOS and also on Desktop/amd64 + llvmfullaot: https://github.com/lewurm/mono/commit/2139a8bd2259c2cfd8e59dc5eba5a68ecdd05ba1

run:
```console
$ make -C mono/mini llvmfullaotcheck
 ```
alas `BinaryFormatter.Serialize` is not exactly a small repro, but it's a start...

**Comment by vargaz** [comment]
Made a different fix.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lewurm** [comment]
The repro still fails on iOS 😕(but not on Desktop/amd64 anymore).

**Comment by lewurm** [comment]
Sorry, I seem to have messed up something when cleaning the repro on iOS (clean != `rm -rf obj bin` 😞 ). With the updated PR the mscorlib tests also passes on iOS64+Release.

File: mono/metadata/class-init.c [hunk 1]
```diff
@@ -1266,7 +1266,15 @@ make_generic_param_class (MonoGenericParam *param)
 	 * This makes sure the the value size of this class is equal to the size of the types the gparam is
 	 * constrained to, the JIT depends on this.
 	 */
-	klass->instance_size = MONO_ABI_SIZEOF (MonoObject) + mono_type_stack_size_internal (m_class_get_byval_arg (klass), NULL, TRUE);
```
Comment (BrzVlad) [inline mono/metadata/class-init.c]: Isn't the real problem that we use `mono_type_stack_size_internal`, which should be used only to compute storage on stack and not on class layout ? And we shouldn't use it anymore at all here ? Wouldn't using `mono_type_size` just work ?

Comment (vargaz) [reply mono/metadata/class-init.c]: That might work as well, there are so many size functions.

Comment (BrzVlad) [reply mono/metadata/class-init.c]: @vargaz  Added a commit on your branch which also fixes the bugs and it looks cleaner. Let me know if it looks OK, so we can merge it for the products once the tests are finished.

Comment (vargaz) [reply mono/metadata/class-init.c]: Looks ok.

**Comment by BrzVlad** [comment]
@monojenkins backport 2019-12

**Pull request merged**